src/test/java/com/amazonaws/services/kinesis/clientlibrary/lib/worker/ShardSyncTaskManagerTest.java

```diff
@@ -0,0 +1,126 @@
+import static org.mockito.Matchers.anyBoolean;
+import static org.mockito.Mockito.mock;
+import static org.mockito.Mockito.times;
+import static org.mockito.Mockito.verify;
+import static org.mockito.Mockito.when;
+
+public class ShardSyncTaskManagerTest {
+    private static final long SHARDSYNCER_INVOCATION_WAIT_SLEEP_MILLIS = 100;
+    private static final InitialPositionInStreamExtended INITIAL_POSITION_IN_STREAM_EXTENDED = InitialPositionInStreamExtended
+            .newInitialPosition(InitialPositionInStream.TRIM_HORIZON);
+    private static final boolean CLEANUP_LEASES_SHARD_COMPLETION = Boolean.TRUE;
+    private static final boolean IGNORE_UNEXPECTED_CHILD_SHARDS = Boolean.TRUE;
+    private static final long SHARD_SYNC_IDLE_TIME_MILLIS = 0;
+
+    @Mock private IKinesisProxy mockKinesisProxy;
+    @Mock private ILeaseManager<KinesisClientLease> mockLeaseManager;
+    @Mock private IMetricsFactory mockMetricsFactory;
+    @Mock private IMetricsScope mockMetricsScope;
+
+    private ShardSyncTaskManager shardSyncTaskManager;
+    private ShardSyncer pausableNoOpShardSyncer;
+    private ShardSyncer mockShardSyncer;
+    private CountDownLatch countDownLatch;
+
+    @Before public void setup() {
+        MockitoAnnotations.initMocks(this);
+        when(mockMetricsFactory.createMetrics()).thenReturn(mockMetricsScope);
+        countDownLatch = new CountDownLatch(1);
+        pausableNoOpShardSyncer = new PausableNoOpShardSyncer(countDownLatch);
+        mockShardSyncer = mock(ShardSyncer.class, delegatesTo(pausableNoOpShardSyncer));
+        shardSyncTaskManager = new ShardSyncTaskManager(mockKinesisProxy, mockLeaseManager,
+                INITIAL_POSITION_IN_STREAM_EXTENDED, CLEANUP_LEASES_SHARD_COMPLETION, IGNORE_UNEXPECTED_CHILD_SHARDS,
+                SHARD_SYNC_IDLE_TIME_MILLIS, mockMetricsFactory, Executors.newSingleThreadExecutor(), mockShardSyncer);
+    }
+
+    @Test public void testShardSyncIdempotency() throws Exception {
+        shardSyncTaskManager.syncShardAndLeaseInfo(new ArrayList<>());
+        Thread.sleep(SHARDSYNCER_INVOCATION_WAIT_SLEEP_MILLIS); // small pause to wait for shardSyncer invocations.
+        verify(mockShardSyncer, times(1))
+                .checkAndCreateLeasesForNewShards(Matchers.any(), Matchers.any(), Matchers.any(), anyBoolean(),
+                        anyBoolean(), Matchers.any());
+        // Invoke a few more times. This would flip shardSyncRequestPending to true in ShardSyncTaskManager.
+        int count = 0;
+        while (count++ < 5) {
+            shardSyncTaskManager.syncShardAndLeaseInfo(new ArrayList<>());
+        }
+        // Since countDownLatch is still blocked, previous ShardSyncTask is still running, hence no new invocations.
+        verify(mockShardSyncer, times(1))
+                .checkAndCreateLeasesForNewShards(Matchers.any(), Matchers.any(), Matchers.any(), anyBoolean(),
+                        anyBoolean(), Matchers.any());
```

Comment:
Can we gracefully exit by counting down the latch?

Reply:
Added.